processor/src/main/java/org/mapstruct/ap/internal/model/common/Type.java

```diff
@@ -1121,59 +1200,237 @@ public boolean isLiteral() {
+                    if ( Object.class.getName().equals( declaredSuperType.getFullyQualifiedName() ) ) {
+                        continue;
+                    }
+                    ResolvedPair result = visitDeclared( parameterized, declaredSuperType );
+                    if ( result != super.DEFAULT_VALUE  ) {
+                        results.add( result );
+                    }
+                }
+
+                for ( TypeMirror parameterizedSuper : types.directSupertypes( parameterized ) ) {
+                    if ( isJavaLangObject( parameterizedSuper ) ) {
+                        continue;
+                    }
+                    ResolvedPair result = visitDeclared( (DeclaredType) parameterizedSuper, declared );
+                    if ( result != super.DEFAULT_VALUE  ) {
+                        results.add( result );
+                    }
+                }
+            }
+            if ( results.isEmpty() ) {
+                return super.DEFAULT_VALUE;
+            }
+            else {
+                return results.stream().allMatch( results.get( 0 )::equals ) ? results.get( 0 ) : super.DEFAULT_VALUE;
+            }
         }
+
+        private boolean isJavaLangObject(TypeMirror type) {
+            if ( type instanceof DeclaredType ) {
+                return ( (TypeElement) ( (DeclaredType) type ).asElement() ).getQualifiedName()
+                                                                            .contentEquals( Object.class.getName() );
+            }
+            return false;
+        }
+    }
+
+    /**
+     * Reflects any Resolved Pair, examples are
+     * T, String
+     * ? extends T, BigDecimal
+     * T[], Integer[]
+     */
+    public static class ResolvedPair {
+
+        public ResolvedPair(Type parameter, Type match) {
+            this.parameter = parameter;
+            this.match = match;
+        }
+
+        /**
+         * parameter, e.g. T, ? extends T or T[]
+         */
+        private Type parameter;
+
+        /**
+         * match, e.g. String, BigDecimal, Integer[]
+         */
+        private Type match;
+
+        public Type getParameter() {
+            return parameter;
+        }
+
+        public Type getMatch() {
+            return match;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if ( this == o ) {
+                return true;
+            }
+            if ( o == null || getClass() != o.getClass() ) {
+                return false;
+            }
+            ResolvedPair that = (ResolvedPair) o;
+            return parameter.equals( that.parameter );
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash( parameter );
+        }
+    }
+
+    /**
+     * Gets the boxed equivalent type if the type is primitive, int will return Integer
+     *
+     * @return boxed equivalent
+     */
+    public Type getBoxedEquivalent() {
```

Comment:
the method's body: `typeUtils.boxedClass( (PrimitiveType) typeMirror )` So give it a primitive Type and it will find it's boxed equivalent (mostly used for comparison purposes). .But perhaps I don't get your comment..

Reply:
Looking at it now it makes sense. No idea what I was thinking then :).